pkg/controller/gcppubsub/reconcile.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcppubsub
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/knative/eventing-sources/pkg/controller/sinks"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/knative/eventing-sources/pkg/apis/sources/v1alpha1"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	finalizerName = controllerAgentName
+)
+
+var (
+	trueVal = true
+)
+
+type reconciler struct {
+	client        client.Client
+	dynamicClient dynamic.Interface
+
+	receiveAdapterImage              string
+	receiveAdapterServiceAccountName string
+}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) InjectConfig(c *rest.Config) error {
+	var err error
+	r.dynamicClient, err = dynamic.NewForConfig(c)
+	return err
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) (runtime.Object, error) {
+	logger := logging.FromContext(ctx).Desugar()
+
+	src, ok := object.(*v1alpha1.GcpPubSubSource)
+	if !ok {
+		logger.Error("could not find GcpPubSub source", zap.Any("object", object))
+		return object, nil
+	}
+
+	// This Source attempts to reconcile three things.
+	// 1. Determine the sink's URI.
+	//     - Nothing to delete.
+	// 2. Create a receive adapter in the form of a Deployment.
+	//     - Will be garbage collected by K8s when this GcpPubSubSource is deleted.
+	// 3. Register that receive adapter as a Push endpoint for the specified GCP PubSub Topic.
```

Comment:
I thought this was a Pull based Receive Adapter?

Reply:
You are right. Updated the comment.

(Originally I was working on push based, which is where the comment originated)